File: alwin-middleware-grapescode/alwin-core-api/src/main/java/com/codersteam/alwin/core/api/service/issue/IssueTerminationService.java
package com.codersteam.alwin.core.api.service.issue;

import com.codersteam.alwin.core.api.model.issue.IssueTerminationRequestDto;

import javax.ejb.Local;

/**
 * Serwis do obsługi żądań przedterminowego zakończenia zlecenia
 *
 * @author Piotr Naroznik
 */
@Local
public interface IssueTerminationService {

    /**
     * Zaakceptowanie żądania o przedterminowe zakończenie zlecenia oraz zakończenie zlecenia.
     *
     * @param issueTerminationRequestId - identyfikator żądania o przedterminowe zakończenie zlecenia
     * @param operatorId                - identyfikator operatora
     * @param terminationRequest        - dane do zakończenia zlecenia
     */
    void accept(Long issueTerminationRequestId, Long operatorId, IssueTerminationRequestDto terminationRequest);

    /**
     * Odrzucenie żądania o przedterminowe zakończenie zlecenia.
     *
     * @param issueTerminationRequestId - identyfikator żądania o przedterminowe zakończenie zlecenia
     * @param operatorId                - identyfikator operatora
     * @param terminationRequest        - dane do zakończenia zlecenia
     */
    void reject(Long issueTerminationRequestId, Long operatorId, IssueTerminationRequestDto terminationRequest);

    /**
     * Pobieranie otwartego żądania o przedterminowe zakończenia zlecenia
     *
     * @param issueId - identyfikator zlecenia
     * @return żądanie przedterminowego zakończenie zlecenia
     */
    IssueTerminationRequestDto findTerminationRequestByIssueId(Long issueId);

    /**
     * Pobieranie żądania o przedterminowe zakończenia zlecenia
     *
     * @param issueTerminationRequestId - identyfikator zlecenia
     * @return żądanie przedterminowego zakończenie zlecenia
     */
    IssueTerminationRequestDto findTerminationRequestById(Long issueTerminationRequestId);
}

File: alwin-web-grapescode/src/app/issues/shared/issue-termination.service.ts
import {AlwinHttpService} from '../../shared/authentication/alwin-http.service';
import {Injectable} from '@angular/core';
import {IssueTerminationRequest} from './terminate-issue-request';
import {Observable} from 'rxjs';
import {IssueTermination} from './issue-termination';

@Injectable()
export class IssueTerminationService {

  constructor(private http: AlwinHttpService) {
  }

  /**
   * Przedterminowe zakończenie zlecenia windykacyjnego
   *
   * @param issueId - identyfikator zlecenia
   * @param issueTerminationRequest - dane do zamknięcia zlecenia
   * @return komunikat: zlecenie zakończone lub żądanie o przedterminowe zakończenie zlecenia zostało utowrzone
   */
  terminateIssue(issueId: number, issueTerminationRequest: IssueTerminationRequest): Observable<string> {
    return this.http.post(`issues/terminate/${issueId}`, issueTerminationRequest);
  }

  /**
   * Pobieranie otwartego żądania o przedterminowe zakończenie zlecenia
   *
   * @param issueId - identyfikator zlecenia
   * @return żądanie o przedterminowe zakończenie zlecenia
   */
  findTerminationRequest(issueId: number): Observable<IssueTermination> {
    return this.http.get(`issues/termination/${issueId}`);
  }

  /**
   * Zaakceptowanie żądania o przedterminowe zakończenie zlecenia oraz zakończenie zlecenia
   *
   * @param terminationRequestId - identyfikator żądania o przedterminowe zakończenie zlecenia
   * @param terminationRequest   - dane do zaakcpetowania żądania
   * @return pusta odpowiedź ze statusem 200
   */
  acceptIssueTermination(terminationRequestId: number, terminationRequest: IssueTermination): any {
    return this.http.postObserve(`issues/termination/accept/${terminationRequestId}`, terminationRequest);
  }

  /**
   * Odrzucenie żądania o przedterminowe zakończenie zlecenia
   *
   * @param terminationRequestId - identyfikator żądania o przedterminowe zakończenie zlecenia
   * @param terminationRequest   - dane do odrzucenia żądania
   * @return pusta odpowiedź ze statusem 200
   */
  rejectIssueTermination(terminationRequestId: number, terminationRequest: IssueTermination): any {
    return this.http.postObserve(`issues/termination/reject/${terminationRequestId}`, terminationRequest);
  }
}
